# Capstone — Which page do you fix first?

**John Andrei Martinez · FlyRank ML Internship · Lane 2, Refresh / Content Opportunity Scoring**

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Zeref538/flyrank-ml-internship/blob/main/work/notebooks/capstone.ipynb)

This notebook mirrors the deployed paper section for section, and produces the
numbers it reports. If a figure here disagrees with the paper, this notebook wins
and the paper is stale.

**Paper:** https://zeref538.github.io/flyrank-ml-internship/


## 1. Question

**Lane 2 — Refresh / Content Opportunity Scoring.**

A FlyRank client publishes more content than anyone can maintain. Pages age,
search shifts underneath them, and traffic drifts down slowly enough that nobody
notices a specific page dying. The binding constraint is not knowledge, it is
hours: a strategist has a fixed number of review slots each month against
thousands of candidate pages.

> **Given a fixed number of review hours, which pages should a content team open
> first?**

What that question deliberately does *not* ask: whether refreshing a page will
recover its traffic. Nothing here can answer that, because nobody ran an
experiment assigning pages to be refreshed. This is decision support — a better
order to work through — and never a promise about an outcome.

## 2. Data

Starter release `data/raw/content_refresh_anonymized.csv` — 30,000 rows, 44
columns, one row per pseudonymised content item, 32 clients.

**Date windows.** There is no calendar date in this snapshot — no `report_date`.
Every window is relative to when the extract was taken: `*_90d` trailing totals,
plus a `*_last_30d` / `*_prev_30d` pair splitting the most recent sixty days.
That is why nothing here can be re-run "as of" an earlier date.

In [1]:
import numpy as np, pandas as pd, json
from pathlib import Path

OUT = Path("../outputs"); OUT.mkdir(parents=True, exist_ok=True)
FIG = Path("../figures"); FIG.mkdir(parents=True, exist_ok=True)

df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")
df["is_declining"] = df["trend_direction"].str.lower().eq("down").astype(int)

print(f"rows: {len(df):,} | columns: {df.shape[1]} | clients: {df.client_id.nunique()}")
print(f"content_age_days      {int(df.content_age_days.min())}-{int(df.content_age_days.max())}")
print(f"days_since_last_update {int(df.days_since_last_update.min())}-{int(df.days_since_last_update.max())}")
print(f"declining rate (all rows): {df.is_declining.mean():.3f}")

# --- eligibility gate, fixed before any modelling, applied to rule and models alike
elig = df[(df.impressions_90d >= 250) & (df.avg_position > 0) & (df.avg_position <= 20)].copy()
print()
print("gate: impressions_90d >= 250, avg_position in (0, 20]")
print("  avg_position == 0 means 'no data' in this dataset, not rank zero:",
      int((df.avg_position == 0).sum()), "rows")
print(f"  eligible: {len(elig):,} pages across {elig.client_id.nunique()} clients")
print(f"  declining rate among eligible: {elig.is_declining.mean():.3f}")
print()
print("This gate is a survivorship filter and is named as one: everything below")
print("describes pages that ALREADY get meaningful search traffic.")

rows: 30,000 | columns: 45 | clients: 32
content_age_days      90-564
days_since_last_update 1-373
declining rate (all rows): 0.542

gate: impressions_90d >= 250, avg_position in (0, 20]
  avg_position == 0 means 'no data' in this dataset, not rank zero: 1205 rows
  eligible: 13,562 pages across 29 clients
  declining rate among eligible: 0.610

This gate is a survivorship filter and is named as one: everything below
describes pages that ALREADY get meaningful search traffic.


## 3. Methodology

Assumptions, features, label, baseline, validation design, leakage checks.

In [2]:
# --- label -----------------------------------------------------------------
# is_declining <- trend_direction <- trend_pct. Both are banned as features.
BANNED = {"trend_direction", "trend_pct", "is_declining"}

# --- the baseline: a rule a strategist can read in ten seconds --------------
# Written BEFORE any model, and kept as the thing to beat. Two terms: how far a
# page's click-through rate sits below the median for its own position tier, and
# how long since it was last updated.
tier_med = elig.groupby("position_tier")["ctr"].transform("median")
elig["tier_median_ctr"] = tier_med
elig["ctr_gap_ratio"] = ((tier_med - elig.ctr) / tier_med.replace(0, np.nan)).clip(lower=0).fillna(0)
elig["freshness_term"] = (elig.days_since_last_update / 365).clip(upper=1)
elig["rule_score"] = elig.ctr_gap_ratio + elig.freshness_term

FEATURES = ["ctr_gap_ratio", "freshness_term", "ctr", "avg_position", "impressions_90d",
            "clicks_90d", "days_since_last_update", "content_age_days", "word_count",
            "engagement_rate", "scroll_rate", "days_with_impressions", "search_volume",
            "competition"]
assert not (set(FEATURES) & BANNED), "a label-derived column got into the feature list"

X = elig[FEATURES].replace([np.inf, -np.inf], np.nan).fillna(0)
y = elig.is_declining.values

print("baseline rule:  score = ctr_gap_ratio + freshness_term")
print(f"features: {len(FEATURES)}  |  banned as features: {sorted(BANNED)}")
print()
print("The rule's blind spot is visible in its own arithmetic: a page with zero clicks")
print("gets the maximum possible ctr_gap_ratio, so pages nobody clicks float to the top.")
print("That comes back in section 4.")

baseline rule:  score = ctr_gap_ratio + freshness_term
features: 14  |  banned as features: ['is_declining', 'trend_direction', 'trend_pct']

The rule's blind spot is visible in its own arithmetic: a page with zero clicks
gets the maximum possible ctr_gap_ratio, so pages nobody clicks float to the top.
That comes back in section 4.


In [3]:
# --- validation design: the decision that mattered most ---------------------
# Pages from one client share a CMS, an audience and a tracking setup. Let some of
# a client's pages into training and the rest into test, and a model can learn
# "this is client 7, and client 7 declines" -- which is not a fact about content.
from sklearn.model_selection import GroupShuffleSplit, ShuffleSplit

tr, te = next(GroupShuffleSplit(n_splits=1, test_size=0.3, random_state=42)
              .split(X, y, groups=elig.client_id))
assert not (set(elig.iloc[tr].client_id) & set(elig.iloc[te].client_id)), \
    "clients leaked across the split"

BASE_RATE = float(y[te].mean())
print(f"train {len(tr):,} pages / {elig.iloc[tr].client_id.nunique()} clients")
print(f"test  {len(te):,} pages / {elig.iloc[te].client_id.nunique()} clients")
print(f"clients on both sides: 0  (asserted above, not assumed)")
print(f"test base rate: {BASE_RATE:.3f}  <- the number every precision below is judged against")


def precision_at_k(scores, labels, k):
    "Share of the top k ranked items that really were declining."
    order = np.argsort(-np.asarray(scores), kind="stable")
    return float(np.asarray(labels)[order[:k]].mean())


# metric self-check runs before anything depends on it
assert precision_at_k([3, 2, 1], [1, 0, 1], 2) == 0.5, "metric is wrong, stop here"
print("\nmetric self-check passed")

train 10,607 pages / 20 clients
test  2,955 pages / 9 clients
clients on both sides: 0  (asserted above, not assumed)
test base rate: 0.551  <- the number every precision below is judged against

metric self-check passed


In [4]:
# --- leakage check, part 1: break the alarm on purpose ---------------------
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import roc_auc_score

def auc_with(extra_cols):
    cols = FEATURES + extra_cols
    Xa = elig[cols].replace([np.inf, -np.inf], np.nan).fillna(0)
    m = GradientBoostingClassifier(random_state=42).fit(Xa.iloc[tr], y[tr])
    s = m.predict_proba(Xa.iloc[te])[:, 1]
    imp = pd.Series(m.feature_importances_, index=cols).sort_values(ascending=False)
    return roc_auc_score(y[te], s), imp

auc_clean, imp_clean = auc_with([])
auc_leak,  imp_leak  = auc_with(["trend_pct"])

print(f"clean features      ROC-AUC {auc_clean:.3f}   top: {imp_clean.index[0]} ({imp_clean.iloc[0]:.3f})")
print(f"+ trend_pct         ROC-AUC {auc_leak:.3f}   top: {imp_leak.index[0]} ({imp_leak.iloc[0]:.3f})")
print()
print("A jump to a perfect score with the leaked column taking ~all the importance is the")
print("confession we wanted. The harness detects leakage, which is what makes the clean")
print("number trustworthy as a measurement rather than a hope.")

clean features      ROC-AUC 0.618   top: content_age_days (0.187)
+ trend_pct         ROC-AUC 1.000   top: trend_pct (0.999)

A jump to a perfect score with the leaked column taking ~all the importance is the
confession we wanted. The harness detects leakage, which is what makes the clean
number trustworthy as a measurement rather than a hope.


In [5]:
# --- leakage check, part 2: is there ANOTHER way to reach the label? -------
# The guidance names two columns to avoid. But a label is computed from something,
# and those ingredients usually sit in the same table under friendlier names.
# trend_pct is a percentage change over 30-day windows -- and those windows are here.
for base in ["clicks", "impressions", "sessions"]:
    a, b = df[f"{base}_last_30d"], df[f"{base}_prev_30d"]
    rebuilt = ((a - b) / b.replace(0, np.nan)) * 100
    m = df.trend_pct.notna() & rebuilt.notna()
    corr = rebuilt[m].corr(df.trend_pct[m])
    match = np.isclose(rebuilt[m], df.trend_pct[m], atol=0.6).mean()
    print(f"{base:12} corr={corr:.4f}   within 0.6pp: {match:6.1%}   (n={int(m.sum()):,})")

print()
print("trend_pct == (impressions_last_30d - impressions_prev_30d) / impressions_prev_30d * 100")
print("Exact. Neither column is on any ban list, and neither name suggests danger.")

auc_hidden, imp_hidden = auc_with(["impressions_last_30d", "impressions_prev_30d"])
print()
print(f"clean                          ROC-AUC {auc_clean:.3f}")
print(f"+ the two 30-day impressions   ROC-AUC {auc_hidden:.3f}   top: {imp_hidden.index[0]} ({imp_hidden.iloc[0]:.3f})")
print(f"cost of the hidden leak: {auc_hidden - auc_clean:+.3f} AUC")
print()
print("Larger than the split-design effect (+0.080), and harder to catch, because the")
print("split problem at least has a name people teach. Lesson: banning the label is not")
print("enough -- trace how the label was computed and ban its ingredients too.")
print()
print("Everything from here on uses the clean feature set.")

clicks       corr=0.0913   within 0.6pp:   1.0%   (n=11,759)
impressions  corr=1.0000   within 0.6pp: 100.0%   (n=26,612)
sessions     corr=0.0520   within 0.6pp:   3.8%   (n=21,873)

trend_pct == (impressions_last_30d - impressions_prev_30d) / impressions_prev_30d * 100
Exact. Neither column is on any ban list, and neither name suggests danger.



clean                          ROC-AUC 0.618
+ the two 30-day impressions   ROC-AUC 0.992   top: impressions_last_30d (0.579)
cost of the hidden leak: +0.374 AUC

Larger than the split-design effect (+0.080), and harder to catch, because the
split problem at least has a name people teach. Lesson: banning the label is not
enough -- trace how the label was computed and ban its ingredients too.

Everything from here on uses the clean feature set.


## 4. Results (vs baseline)

Every method scored on the same held-out clients, against the same base rate.

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import average_precision_score

models = {
    "logistic_regression": make_pipeline(StandardScaler(),
                                         LogisticRegression(max_iter=2000, class_weight="balanced")),
    "decision_tree": DecisionTreeClassifier(max_depth=4, class_weight="balanced", random_state=42),
    "random_forest": RandomForestClassifier(n_estimators=300, class_weight="balanced",
                                            random_state=42, n_jobs=-1),
    "gradient_boosting": GradientBoostingClassifier(random_state=42),
}

y_te = y[te]
rows = [dict(method="base rate (random order)", P_at_10=round(BASE_RATE, 3),
             P_at_20=round(BASE_RATE, 3), P_at_50=round(BASE_RATE, 3), ROC_AUC=0.500)]

rule_te = elig.rule_score.values[te]
rows.append(dict(method="hand-written rule",
                 P_at_10=round(precision_at_k(rule_te, y_te, 10), 3),
                 P_at_20=round(precision_at_k(rule_te, y_te, 20), 3),
                 P_at_50=round(precision_at_k(rule_te, y_te, 50), 3),
                 ROC_AUC=round(roc_auc_score(y_te, rule_te), 3)))

scores = {}
for name, m in models.items():
    m.fit(X.iloc[tr], y[tr])
    s = m.predict_proba(X.iloc[te])[:, 1]
    scores[name] = s
    rows.append(dict(method=name,
                     P_at_10=round(precision_at_k(s, y_te, 10), 3),
                     P_at_20=round(precision_at_k(s, y_te, 20), 3),
                     P_at_50=round(precision_at_k(s, y_te, 50), 3),
                     ROC_AUC=round(roc_auc_score(y_te, s), 3)))

results = pd.DataFrame(rows)
print(results.to_string(index=False))

                  method  P_at_10  P_at_20  P_at_50  ROC_AUC
base rate (random order)    0.551    0.551    0.551    0.500
       hand-written rule    0.900    0.950    0.860    0.578
     logistic_regression    0.800    0.800    0.780    0.639
           decision_tree    0.700    0.500    0.500    0.616
           random_forest    0.700    0.700    0.780    0.623
       gradient_boosting    0.900    0.900    0.880    0.618


In [7]:
# --- how big is a "win" here, really? --------------------------------------
# Before celebrating any gap, measure how much the rule's own score moves when
# nothing changes except the order of tied pages. Many pages share a rule score,
# and argsort breaks those ties by position in the file -- an arbitrary choice.
rng = np.random.default_rng(0)
tie_runs = []
for _ in range(200):
    jitter = rng.permutation(len(rule_te))          # random tie-break, same scores
    order = np.lexsort((jitter, -rule_te))
    tie_runs.append(float(y_te[order[:50]].mean()))

lo, hi = min(tie_runs), max(tie_runs)
gb_p50 = results.loc[results.method == "gradient_boosting", "P_at_50"].iloc[0]
rule_p50 = results.loc[results.method == "hand-written rule", "P_at_50"].iloc[0]

print(f"rule P@50 across 200 random tie-breaks: {lo:.2f} to {hi:.2f}")
print(f"gradient boosting {gb_p50:.2f}  vs  rule {rule_p50:.2f}   gap {gb_p50 - rule_p50:+.2f}")
print()
print(f"A {abs(gb_p50 - rule_p50):.2f} gap inside a {hi - lo:.2f}-wide noise band is not a win.")
print("HONEST CLAIM: the model did not demonstrably beat a five-line rule.")
print("A team should ship the rule -- it explains itself and costs nothing to maintain.")

rule P@50 across 200 random tie-breaks: 0.76 to 0.90
gradient boosting 0.88  vs  rule 0.86   gap +0.02

A 0.02 gap inside a 0.14-wide noise band is not a win.
HONEST CLAIM: the model did not demonstrably beat a five-line rule.
A team should ship the rule -- it explains itself and costs nothing to maintain.


In [8]:
# --- the validation finding, and why v1 of this notebook disagreed ---------
tr_r, te_r = next(ShuffleSplit(n_splits=1, test_size=0.3, random_state=42).split(X, y))
m_r = GradientBoostingClassifier(random_state=42).fit(X.iloc[tr_r], y[tr_r])
auc_random = roc_auc_score(y[te_r], m_r.predict_proba(X.iloc[te_r])[:, 1])
shared = len(set(elig.iloc[tr_r].client_id) & set(elig.iloc[te_r].client_id))

print(f"random row split   ROC-AUC {auc_random:.3f}   clients on both sides: {shared}")
print(f"grouped by client  ROC-AUC {auc_clean:.3f}   clients on both sides: 0")
print(f"gap {auc_random - auc_clean:+.3f}")
print()
print("Bigger than the difference between any two models above.")
print()
print("--- why an earlier version of this notebook reported different numbers ---")
print("v1 ran on all 30,000 rows with no eligibility gate, and used a different baseline")
print("('stale 180d+ AND visible 500+', ranked by impressions). It reported the rule at")
print("P@50 0.64. That is not a contradiction, it is a different population and a")
print("different rule. The gated version is the one reported, because scoring pages with")
print("almost no impressions means ranking noise, and the CTR-gap rule is the stronger")
print("baseline -- which makes it the harder one for a model to beat, which is the point.")

random row split   ROC-AUC 0.728   clients on both sides: 27
grouped by client  ROC-AUC 0.618   clients on both sides: 0
gap +0.110

Bigger than the difference between any two models above.

--- why an earlier version of this notebook reported different numbers ---
v1 ran on all 30,000 rows with no eligibility gate, and used a different baseline
('stale 180d+ AND visible 500+', ranked by impressions). It reported the rule at
P@50 0.64. That is not a contradiction, it is a different population and a
different rule. The gated version is the one reported, because scoring pages with
almost no impressions means ranking noise, and the CTR-gap rule is the stronger
baseline -- which makes it the harder one for a model to beat, which is the point.


In [9]:
# --- how much of this rests on nine clients? -------------------------------
# Precision@50 is 44 pages out of 50, drawn from 9 clients. The tie-break band above
# shows the RULE is unstable; it says nothing about how uncertain the MODEL's 0.88 is.
# So: a cluster bootstrap. Resample whole CLIENTS with replacement, not pages, because
# pages inside a client share a CMS, an audience and a tracking setup -- treating them
# as independent would give a falsely narrow interval.
t_eval = elig.iloc[te].copy()
t_eval["model"] = scores["gradient_boosting"]
t_eval["rule"]  = elig.rule_score.values[te]
t_eval["y"]     = y_te

def p50(fr, col):
    return float(fr.nlargest(50, col).y.mean())

rng_b = np.random.default_rng(0)
cl = t_eval.client_id.unique()
boot_m, boot_r, boot_d, boot_a = [], [], [], []
for _ in range(2000):
    pick = rng_b.choice(cl, size=len(cl), replace=True)
    b = pd.concat([t_eval[t_eval.client_id == c] for c in pick], ignore_index=True)
    if b.y.nunique() < 2:
        continue
    m_, r_ = p50(b, "model"), p50(b, "rule")
    boot_m.append(m_); boot_r.append(r_); boot_d.append(m_ - r_)
    boot_a.append(roc_auc_score(b.y, b.model))

def ci(v):
    v = np.array(v); return float(np.percentile(v, 2.5)), float(np.percentile(v, 97.5))

CI_M, CI_R, CI_D, CI_A = ci(boot_m), ci(boot_r), ci(boot_d), ci(boot_a)
BOOT_AHEAD = float((np.array(boot_d) > 0).mean())

print(f"cluster bootstrap, {len(boot_m)} resamples of {len(cl)} held-out clients")
print()
print(f"  model P@50    {p50(t_eval,'model'):.3f}   95% CI [{CI_M[0]:.3f}, {CI_M[1]:.3f}]")
print(f"  rule  P@50    {p50(t_eval,'rule'):.3f}   95% CI [{CI_R[0]:.3f}, {CI_R[1]:.3f}]")
print(f"  model - rule  {p50(t_eval,'model')-p50(t_eval,'rule'):+.3f}   95% CI [{CI_D[0]:+.3f}, {CI_D[1]:+.3f}]")
print(f"  model ROC-AUC {auc_clean:.3f}   95% CI [{CI_A[0]:.3f}, {CI_A[1]:.3f}]")
print()
print(f"  model ahead in {BOOT_AHEAD:.0%} of resamples")
print(f"  interval on the difference includes zero: "
      f"{'YES' if CI_D[0] < 0 < CI_D[1] else 'no'}")
print()
print("Two independent arguments now reach the same place -- the tie-break band and a")
print("standard confidence interval. The model is not distinguishable from the rule.")
print()
print(f"Note the width of the model's own interval ({CI_M[1]-CI_M[0]:.3f}). Nine clients is")
print("not many, and that width is the honest size of what nine clients can tell you.")

cluster bootstrap, 2000 resamples of 9 held-out clients

  model P@50    0.880   95% CI [0.600, 0.920]
  rule  P@50    0.860   95% CI [0.720, 0.900]
  model - rule  +0.020   95% CI [-0.260, +0.140]
  model ROC-AUC 0.618   95% CI [0.562, 0.657]

  model ahead in 61% of resamples
  interval on the difference includes zero: YES

Two independent arguments now reach the same place -- the tie-break band and a
standard confidence interval. The model is not distinguishable from the rule.

Note the width of the model's own interval (0.320). Nine clients is
not many, and that width is the honest size of what nine clients can tell you.


## 5. Limitations

**What this work cannot claim:**

- **Nothing causal.** One snapshot, no intervention, nobody assigned pages to be
  refreshed. "These look worth reviewing first" is the ceiling.
- **Nine held-out clients.** Enough to expose the split effect, too few to
  characterise behaviour across the industries FlyRank serves.
- **The label is the dataset's definition of declining**, not a measure of
  business harm.
- **Survivorship.** The gate keeps pages with 250+ impressions. Everything here
  describes content that already gets found.
- **The leak I found is probably not the only one.** I traced one label back to
  its ingredients. There are 44 columns and I tested the ones the trend
  arithmetic pointed at, not all of them.
- **Content decay is not measurable here** — see the next cell.
- **Effect sizes are modest.** ROC-AUC 0.618 against 0.5 for a coin flip.

In [10]:
# --- does the answer depend on the gate I chose? ---------------------------
# Every number above sits on one threshold: impressions_90d >= 250, fixed before
# any modelling. If the conclusion flips when that moves, it is a fact about the
# threshold, not about the problem. So sweep it.
#
# Averaged over 8 client-grouped splits per gate. Without that, a gate change also
# changes WHICH clients land in the test set, and the two effects are inseparable.
from sklearn.model_selection import GroupShuffleSplit as _GSS

def gate_run(gate, seed):
    e = df[(df.impressions_90d >= gate) & (df.avg_position > 0) & (df.avg_position <= 20)].copy()
    tm = e.groupby("position_tier")["ctr"].transform("median")
    e["ctr_gap_ratio"] = ((tm - e.ctr) / tm.replace(0, np.nan)).clip(lower=0).fillna(0)
    e["freshness_term"] = (e.days_since_last_update / 365).clip(upper=1)
    e["rule_score"] = e.ctr_gap_ratio + e.freshness_term
    Xg = e[FEATURES].replace([np.inf, -np.inf], np.nan).fillna(0)
    yg = e.is_declining.values
    tr_g, te_g = next(_GSS(n_splits=1, test_size=0.3, random_state=seed)
                      .split(Xg, yg, groups=e.client_id))
    mg = GradientBoostingClassifier(random_state=seed).fit(Xg.iloc[tr_g], yg[tr_g])
    sg = mg.predict_proba(Xg.iloc[te_g])[:, 1]
    yt = yg[te_g]
    return dict(eligible=len(e), base=yt.mean(),
                p_model=yt[np.argsort(-sg)[:50]].mean(),
                p_rule=yt[np.argsort(-e.rule_score.values[te_g])[:50]].mean())

gate_rows = []
for gate in (100, 250, 500, 1000, 2000):
    R = pd.DataFrame([gate_run(gate, s) for s in range(8)])
    R["diff"] = R.p_model - R.p_rule
    gate_rows.append(dict(gate=gate, eligible=int(R.eligible.iloc[0]),
                          base_rate=round(R.base.mean(), 3),
                          p_model=round(R.p_model.mean(), 3),
                          p_rule=round(R.p_rule.mean(), 3),
                          diff=round(R["diff"].mean(), 3),
                          rule_wins=int((R["diff"] < 0).sum()),
                          lift_model=round((R.p_model - R.base).mean(), 3)))

GATE = pd.DataFrame(gate_rows)
print("Precision@50, mean over 8 client-grouped splits per gate:")
print(GATE.to_string(index=False))
print()
GATE_MAX_DIFF = float(GATE["diff"].abs().max())
print(f"largest model-minus-rule gap across all five gates: {GATE_MAX_DIFF:+.3f}")
print(f"the rule still wins outright in {GATE.rule_wins.min()}-{GATE.rule_wins.max()} of 8 splits at every gate")
print()
print("Eligible pages nearly halve (15,091 -> 7,730) and the gap never clears 0.10.")
print("The non-result is a property of the problem, not of the threshold I picked.")
print()
print("Separate point, and it matters: both methods beat their own base rate by a")
print("wide margin. The ranking task is real. It is the model's edge OVER THE RULE")
print("that this data will not support.")
GATE.to_json(OUT / "capstone_gate_sensitivity.json", orient="records", indent=2)


Precision@50, mean over 8 client-grouped splits per gate:
 gate  eligible  base_rate  p_model  p_rule  diff  rule_wins  lift_model
  100     15091      0.639    0.867   0.838 0.030          3       0.229
  250     13562      0.632    0.860   0.792 0.067          2       0.228
  500     12023      0.571    0.840   0.752 0.087          1       0.269
 1000     10011      0.562    0.835   0.790 0.045          1       0.273
 2000      7730      0.531    0.768   0.740 0.027          2       0.237

largest model-minus-rule gap across all five gates: +0.087
the rule still wins outright in 1-3 of 8 splits at every gate

Eligible pages nearly halve (15,091 -> 7,730) and the gap never clears 0.10.
The non-result is a property of the problem, not of the threshold I picked.

Separate point, and it matters: both methods beat their own base rate by a
wide margin. The ranking task is real. It is the model's edge OVER THE RULE
that this data will not support.


In [11]:
# --- the decay question I could not answer --------------------------------
vc = elig.days_since_last_update.value_counts()
per_client = elig.groupby("client_id").days_since_last_update.nunique()

print("I wanted to show: the longer since an update, the more likely a page is declining.")
print("Then I looked at how the staleness column is actually distributed.\n")
print(f"top 2 values cover {vc.head(2).sum() / len(elig):.0%} of eligible pages")
print(f"distinct values per client -- min {per_client.min()}, median {per_client.median():.0f}, max {per_client.max()}")
print(f"largest client: {int(elig.client_id.value_counts().iloc[0]):,} pages, "
      f"{int(per_client.loc[elig.client_id.value_counts().index[0]])} distinct values")
print(f"range: {int(elig.days_since_last_update.min())}-{int(elig.days_since_last_update.max())} days\n")
print("A per-page 'last edited' date does not look like this. It behaves like a per-client")
print("crawl timestamp, so comparing pages on it mostly compares which client they belong to.")
print("NEGATIVE RESULT: content decay is not measurable in this snapshot.")
print()
print("Knock-on effect worth naming: freshness_term -- half of my own baseline rule -- is")
print("built on this column, so under a client-grouped split it carries almost no signal.")

I wanted to show: the longer since an update, the more likely a page is declining.
Then I looked at how the staleness column is actually distributed.

top 2 values cover 70% of eligible pages
distinct values per client -- min 1, median 3, max 10
largest client: 4,760 pages, 3 distinct values
range: 4-301 days

A per-page 'last edited' date does not look like this. It behaves like a per-client
crawl timestamp, so comparing pages on it mostly compares which client they belong to.
NEGATIVE RESULT: content decay is not measurable in this snapshot.

Knock-on effect worth naming: freshness_term -- half of my own baseline rule -- is
built on this column, so under a client-grouped split it carries almost no signal.


## 6. Ranked recommendations

The queue, with an archetype and a reason code per page — because a score of 0.93
is not something a writer can act on. Full playbook, human-review gates,
monitoring triggers and cost/value in
[`w07_action_playbook.ipynb`](w07_action_playbook.ipynb).

In [12]:
q = elig.iloc[te].copy()
q["model_score"] = scores["gradient_boosting"]

WC_P10 = float(elig.word_count.quantile(0.10))

def archetype(r):
    # Caught FIRST so nothing can overrule it: a page with zero clicks has no click
    # trend that can decline, and it gets the MAXIMUM ctr_gap_ratio precisely because
    # it has no clicks. The rule's blind spot, followed faithfully by the model.
    if r.clicks_90d == 0:                                   return "zero_click_ghost"
    if r.ctr_gap_ratio >= 0.5 and r.impressions_90d >= 500:  return "visible_but_unclicked"
    if 11 <= r.avg_position <= 20:                           return "stuck_on_page_two"
    if pd.isna(r.word_count):                                return "length_unknown"
    if r.word_count <= WC_P10:                               return "short_for_its_class"
    return "no_clear_pattern"

ACTIONS = {
    "zero_click_ghost":      "DO NOT REFRESH - check tracking first",
    "visible_but_unclicked": "Rewrite title and meta description",
    "stuck_on_page_two":     "Improve depth and internal links",
    "length_unknown":        "Analyst check - no word count on file",
    "short_for_its_class":   "Expand coverage",
    "no_clear_pattern":      "Manual look, no preset action",
}

q["archetype"] = q.apply(archetype, axis=1)
q["action"] = q.archetype.map(ACTIONS)
queue = q.sort_values("model_score", ascending=False).reset_index(drop=True)
assert not (set(ACTIONS) - set(queue.archetype.unique())), "an archetype rule never fires"

top50 = queue.head(50)
blocked = int((top50.archetype == "zero_click_ghost").sum())
p50 = precision_at_k(queue.model_score.values, queue.is_declining.values, 50)

print(queue.archetype.value_counts().to_string())
print()
print(f"P@50 {p50:.3f} vs base rate {BASE_RATE:.3f}  ->  lift {p50 / BASE_RATE:.2f}x")
print(f"of the top 50, {blocked} are zero-click pages the playbook BLOCKS from refresh work")
print()
print("Reviewing 50 pages from this queue finds about "
      f"{(p50 - BASE_RATE) * 50:.0f} more genuinely declining pages than reviewing 50 at random.")
print("It does NOT say those refreshes recover traffic - that needs an experiment nobody ran.")
print()
print("The cheapest experiment that would fix that: refresh a random half of the queue,")
print("hold the other half, compare after 90 days.")

archetype
no_clear_pattern         1269
stuck_on_page_two         635
zero_click_ghost          388
visible_but_unclicked     362
length_unknown            215
short_for_its_class        86

P@50 0.880 vs base rate 0.551  ->  lift 1.60x
of the top 50, 24 are zero-click pages the playbook BLOCKS from refresh work

Reviewing 50 pages from this queue finds about 16 more genuinely declining pages than reviewing 50 at random.
It does NOT say those refreshes recover traffic - that needs an experiment nobody ran.

The cheapest experiment that would fix that: refresh a random half of the queue,
hold the other half, compare after 90 days.


## 7. Artifacts the paper embeds

In [13]:
metrics = {
    "lane": "2 - Refresh / Content Opportunity Scoring",
    "split": "GroupShuffleSplit on client_id, test_size=0.3, random_state=42",
    "eligible_pages": int(len(elig)),
    "eligible_clients": int(elig.client_id.nunique()),
    "held_out_pages": int(len(te)),
    "held_out_clients": int(elig.iloc[te].client_id.nunique()),
    "base_rate": round(BASE_RATE, 3),
    "results": results.to_dict(orient="records"),
    "roc_auc_clean": round(float(auc_clean), 3),
    "roc_auc_random_split": round(float(auc_random), 3),
    "split_design_cost": round(float(auc_random - auc_clean), 3),
    "roc_auc_with_hidden_leak": round(float(auc_hidden), 3),
    "hidden_leak_cost": round(float(auc_hidden - auc_clean), 3),
    "roc_auc_with_trend_pct": round(float(auc_leak), 3),
    "rule_tie_band_p50": [round(lo, 2), round(hi, 2)],
    "bootstrap": {
        "method": "cluster bootstrap, 2000 resamples of the held-out clients",
        "model_p50_ci95": [round(CI_M[0], 3), round(CI_M[1], 3)],
        "rule_p50_ci95": [round(CI_R[0], 3), round(CI_R[1], 3)],
        "model_minus_rule_ci95": [round(CI_D[0], 3), round(CI_D[1], 3)],
        "model_roc_auc_ci95": [round(CI_A[0], 3), round(CI_A[1], 3)],
        "resamples_model_ahead": round(BOOT_AHEAD, 2),
        "difference_ci_includes_zero": bool(CI_D[0] < 0 < CI_D[1]),
    },
    "archetype_counts": {k: int(v) for k, v in queue.archetype.value_counts().items()},
    "top50_blocked_zero_click": blocked,
}
path = OUT / "capstone_metrics.json"
path.write_text(json.dumps(metrics, indent=2), encoding="utf-8")
results.to_csv(OUT / "capstone_results_table.csv", index=False)

print(f"{path} ({path.stat().st_size:,} bytes)")
print(f"{OUT / 'capstone_results_table.csv'}")
print()
print("Figures the paper embeds are generated by w07_action_playbook.ipynb and the")
print("paper build, and live in work/figures/ and docs/figures/.")

..\outputs\capstone_metrics.json (2,092 bytes)
..\outputs\capstone_results_table.csv

Figures the paper embeds are generated by w07_action_playbook.ipynb and the
paper build, and live in work/figures/ and docs/figures/.


## Self-check

In [14]:
checks = []
def chk(name, ok, detail=""): checks.append((name, bool(ok), detail))

chk("no label-derived column in the feature list", not (set(FEATURES) & BANNED))
chk("no client on both sides of the split",
    not (set(elig.iloc[tr].client_id) & set(elig.iloc[te].client_id)))
chk("every precision is reported next to its base rate", "base_rate" in metrics)
chk("the queue beats the base rate at K=50", p50 > BASE_RATE, f"{p50:.3f} vs {BASE_RATE:.3f}")
chk("the leak test actually detects a leak", auc_leak > 0.99, f"{auc_leak:.3f}")
chk("the hidden-leak finding reproduces", auc_hidden > auc_clean + 0.3,
    f"{auc_clean:.3f} -> {auc_hidden:.3f}")
chk("every archetype rule fires at least once",
    not (set(ACTIONS) - set(queue.archetype.unique())))
chk("the model-vs-rule interval includes zero", CI_D[0] < 0 < CI_D[1],
    f"[{CI_D[0]:+.3f}, {CI_D[1]:+.3f}]")
chk("the honest claim is not overstated",
    abs(gb_p50 - rule_p50) < (hi - lo),
    f"gap {abs(gb_p50 - rule_p50):.2f} < noise band {hi - lo:.2f}")
chk("the non-result survives every eligibility gate tried", GATE_MAX_DIFF < 0.10,
    f"largest gap {GATE_MAX_DIFF:+.3f} across gates 100-2000")
chk("exports written", path.exists() and path.stat().st_size > 0)

for name, ok, detail in checks:
    print(f"  [{'PASS' if ok else 'FAIL'}] {name}" + (f"  ({detail})" if detail else ""))
assert all(ok for _, ok, _ in checks), "a self-check failed"

print(f"\nall {len(checks)} checks passed")

  [PASS] no label-derived column in the feature list
  [PASS] no client on both sides of the split
  [PASS] every precision is reported next to its base rate
  [PASS] the queue beats the base rate at K=50  (0.880 vs 0.551)
  [PASS] the leak test actually detects a leak  (1.000)
  [PASS] the hidden-leak finding reproduces  (0.618 -> 0.992)
  [PASS] every archetype rule fires at least once
  [PASS] the model-vs-rule interval includes zero  ([-0.260, +0.140])
  [PASS] the honest claim is not overstated  (gap 0.02 < noise band 0.14)
  [PASS] the non-result survives every eligibility gate tried  (largest gap +0.087 across gates 100-2000)
  [PASS] exports written

all 11 checks passed


---

**Public-safety statement.** Built on real, pseudonymised FlyRank client data,
treated as confidential. No client name, URL, query or raw row appears in this
notebook, its outputs, the deployed paper, or the public repository. Pseudonymous
identifiers are used only for grouping and splitting, never as features.

Built on the [FlyRank](https://flyrank.ai) ML Internship dataset.